### PR TITLE
MWAOS-49 - Automatically pause GIFs after a certain number of loops

### LINKED ISSUE #15399: Automatically pause GIFs after a certain number of loops
User story: [https://community-daily.mattermost.com/core/pl/nwiooj57tfdouk36mf7b5j58oy](https://community-daily.mattermost.com/core/pl/nwiooj57tfdouk36mf7b5j58oy)

Design and functional notes can be accessed here:
[https://www.figma.com/proto/7psXW8qYfp0LZTBT1ZWvkO/Gif-Controls?node-id=48%3A8884&viewport=683%2C511%2C0.23019954562187195&scaling=min-zoom](https://www.figma.com/proto/7psXW8qYfp0LZTBT1ZWvkO/Gif-Controls?node-id=48%3A8884&viewport=683%2C511%2C0.23019954562187195&scaling=min-zoom)

**Functional notes:**
The Gif in the center channel should run 6 times, and then pause itself with the Gif
button on top of it.

Clicking the button will run the Gif again for 6 times.

Clicking or tapping the Gif anywhere would first play it, if its not already playing.
And if is playing, then a modal will open up with the Gif playing indefinitely.

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-26973

### PR DESCRIPTION
**Pull request by gitstart**
[Demo Video](https://www.loom.com/share/10dd3d61d6d64d02b4799cc19b9ee93e?sid=50828fb8-1d12-4fd2-9726-cf57b4f71dfb)

## Summary
Automatically pause GIFs after a certain number of loops.

[Closes #15399](https://github.com/mattermost/mattermost/issues/15399)

```release-note
Adds an automatically pause to GIFs after it loops 6 times and renders a play button to trigger the GIF to loop again
```

---
This code was written and reviewed by GitStart Community. Growing great engineers, one PR at a time.

### CODE REVIEW

File: webapp/channels/src/components/gif.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            let duration = 0;
+            for (let i = 0; i < arr.length; i++) {
+                if (arr[i] === 0x21 &&
+                arr[i + 1] === 0xF9 &&
+                arr[i + 2] === 0x04 &&
+                arr[i + 7] === 0x00) {
+                    const delay = (arr[i + 5] << 8) | (arr[i + 4] & 0xFF);
+                    duration += delay < 2 ? 10 : delay;
+                }
+            }
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/gif.tsx:44]: Could you please explain/comment what this logic is doing, just by looking at it its unclear whats the bitwise operator is doing in here

File: webapp/channels/src/components/gif.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                        const still = firstGifFrameUrl(img);
+                        if (still) {
+                            setActualStill(still);
+                        }
+                    }
+                });
+            }
+        };
+
+        updateImages();
```
Comment (larkox) [inline webapp/channels/src/components/gif.tsx:88]: Why do we create a function if we are just calling it afterwards? Wouldn't it make more sense to have here the body of the function directly?

File: webapp/channels/src/components/gif.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    const gifLoopCounterInit = useCallback((fileSrc: string) => {
+        const request = new XMLHttpRequest();
+        request.open('GET', fileSrc, true);
+        request.responseType = 'arraybuffer';
+        request.addEventListener('load', async () => {
+            calcDurationGif(request.response).then((duration: any) => {
+                let newCount = gifLoopCount;
+                intervalId.current = setInterval(() => {
+                    newCount++;
+                    setGifLoopCount(newCount);
```
Comment (larkox) [inline webapp/channels/src/components/gif.tsx:116]: Why would we want to cause a re-render on every loop?

File: webapp/channels/src/components/gif.tsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        };
+
+        countGifAnimation();
+    }, [gif]);
+
+    const gifLoopCounterInit = useCallback((fileSrc: string) => {
+        const request = new XMLHttpRequest();
+        request.open('GET', fileSrc, true);
+        request.responseType = 'arraybuffer';
+        request.addEventListener('load', async () => {
```
Comment (larkox) [inline webapp/channels/src/components/gif.tsx:111]: As far as I can tell, this function is not asynchronous (it doesn't return a promise).

File: webapp/channels/src/components/gif.tsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+                let newCount = gifLoopCount;
+                intervalId.current = setInterval(() => {
+                    newCount++;
+                    setGifLoopCount(newCount);
+                }, duration * 1000);
+            });
+        });
+        request.send();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [playing]);
```
Comment (larkox) [inline webapp/channels/src/components/gif.tsx:122]: I don't understand why this depends on playing.

File: webapp/channels/src/components/gif.tsx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        countGifAnimation();
+    }, [gif]);
+
+    const gifLoopCounterInit = useCallback((fileSrc: string) => {
+        const request = new XMLHttpRequest();
+        request.open('GET', fileSrc, true);
+        request.responseType = 'arraybuffer';
+        request.addEventListener('load', async () => {
+            calcDurationGif(request.response).then((duration: any) => {
+                let newCount = gifLoopCount;
```
Comment (larkox) [inline webapp/channels/src/components/gif.tsx:113]: This function is created once, with one particular value of `gifLoopCount` (1). Therefore the `newCount` will never grow.

**Review by larkox — CHANGES_REQUESTED** [review]
I left some comments, but there seems to be still more things to polish in this PR. Let me know if you want to go through them.

File: webapp/channels/src/components/gif.tsx (hunk 2 above)
Comment (devinbinnie) [reply webapp/channels/src/components/gif.tsx:88]: I can see this making sense if the function is `async`, but it doesn't look like it is.

File: webapp/channels/src/components/gif.tsx (hunk 4 above)
Comment (devinbinnie) [reply webapp/channels/src/components/gif.tsx:111]: We could restructure the below function if you wanted to keep this `async`. I think that would be cleaner.

File: webapp/channels/src/components/gif.tsx (hunk 3 above)
Comment (devinbinnie) [reply webapp/channels/src/components/gif.tsx:116]: Agreed, seems like a good case for `useRef` (thanks Daniel for helping me learn the power of `useRef` :D)

**Review by devinbinnie — COMMENTED** [review]
Just a few comments, in principle looks good though :)

**Comment by larkox** [comment]
@gitstart Are you working on this?

**Comment by larkox** [comment]
@gitstart Any news on this? If not, we will close and put back the help wanted ticket as up for grabs.

**Pull request closed without merging**